File: test_mavros/src/setpoint_position_publisher.cpp
#include <ros/ros.h>
#include <geometry_msgs/PoseStamped.h>

int main(int argc, char** argv) {
    ros::init(argc, argv, "setpoint_position_publisher");
    ros::NodeHandle nh;
    
    ros::Publisher setpoint_pub = nh.advertise<geometry_msgs::PoseStamped>(
        "/mavros/setpoint_position/local", 10);

    ros::Rate rate(20.0); // Publishing rate of 20 Hz
    
    ROS_INFO("Setpoint Position publisher node started. Publishing position commands...");

    while (ros::ok()) {
        geometry_msgs::PoseStamped pose_msg;
        pose_msg.header.stamp = ros::Time::now();
        pose_msg.header.frame_id = "map";  // Adjust if necessary

        // Set desired position (x, y, z) in meters
        pose_msg.pose.position.x = 2.0;
        pose_msg.pose.position.y = 5.0;
        pose_msg.pose.position.z = 3.5;

        // Set desired yaw (converted to quaternion)
        double yaw = 1.57; // 90 degrees in radians
        pose_msg.pose.orientation.w = cos(yaw / 2);
        pose_msg.pose.orientation.x = 0.0;
        pose_msg.pose.orientation.y = 0.0;
        pose_msg.pose.orientation.z = sin(yaw / 2);

	ROS_INFO("Publishing Position: [x: %.2f, y: %.2f, z: %.2f] | Yaw: [yaw: %.2f]",
                 pose_msg.pose.position.x, pose_msg.pose.position.y, pose_msg.pose.position.z,
                 yaw);
                 
        setpoint_pub.publish(pose_msg);

        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}

File: test_mavros/scripts/setpoint_position_publisher.py
#!/usr/bin/env python3

import rospy
from geometry_msgs.msg import PoseStamped
import tf.transformations as tf_trans

def publish_setpoint_position():
    rospy.init_node('setpoint_position_publisher', anonymous=True)
    position_pub = rospy.Publisher('/mavros/setpoint_position/local', PoseStamped, queue_size=10)

    rate = rospy.Rate(10)  # 10 Hz

    rospy.loginfo("Publishing position and yaw setpoints to /mavros/setpoint_position/local")

    while not rospy.is_shutdown():
        position_msg = PoseStamped()
        position_msg.header.stamp = rospy.Time.now()

        # Set position (meters)
        position_msg.pose.position.x = 2.0
        position_msg.pose.position.y = 5.0
        position_msg.pose.position.z = 3.5

        # Set yaw (convert from Euler to Quaternion)
        yaw = 1.57  # 90 degrees in radians
        quaternion = tf_trans.quaternion_from_euler(0, 0, yaw)
        position_msg.pose.orientation.x = quaternion[0]
        position_msg.pose.orientation.y = quaternion[1]
        position_msg.pose.orientation.z = quaternion[2]
        position_msg.pose.orientation.w = quaternion[3]

        position_pub.publish(position_msg)
        rospy.loginfo(f"Publishing -> Position: ({position_msg.pose.position.x}, {position_msg.pose.position.y}, {position_msg.pose.position.z}), "
                      f"Yaw: {yaw:.2f} rad")

        rate.sleep()

if __name__ == '__main__':
    try:
        publish_setpoint_position()
    except rospy.ROSInterruptException:
        rospy.logwarn("Setpoint Position publisher interrupted.")

